File: server/packets/io.go
/**
 * Copyright (C) 2026 Ellie Fisher
 *
 * This file is part of the Word Magnets source code. It may be used under the GNU Affero General
 * Public License v3.0.
 *
 * For full terms, see the LICENSE file or visit https://spdx.org/licenses/AGPL-3.0-or-later.html
 */

package packets

import (
	"bytes"
	"encoding/binary"
	"math"
	"strconv"
	"strings"

	"word-magnets/util"
)

/**
 * PacketReader
 */

type PacketReader struct {
	buffer bytes.Buffer
}

func (reader *PacketReader) IsAtEnd() bool {
	return reader.buffer.Len() <= 0
}

func (reader *PacketReader) PeekU8() uint8 {
	var value uint8

	if !reader.IsAtEnd() {
		value = reader.buffer.Bytes()[0]
	}

	return value
}

func (reader *PacketReader) ReadU8() uint8 {
	var value uint8

	if !reader.IsAtEnd() {
		value, _ = reader.buffer.ReadByte()
	}

	return value
}

func (reader *PacketReader) ReadBool() bool {
	return reader.ReadU8() != 0
}

func (reader *PacketReader) ReadString() string {
	length := reader.ReadU8()

	var str strings.Builder
	str.Grow(int(length))

	for i := 0; i < int(length) && !reader.IsAtEnd(); i++ {
		str.WriteByte(reader.ReadU8())
	}

	return str.String()
}

func (reader *PacketReader) MatchU8(value uint8) bool {
	matches := reader.PeekU8() == value

	if matches {
		reader.ReadU8()
	}

	return matches
}

func NewPacketReader(packet []byte) *PacketReader {
	return &PacketReader{*bytes.NewBuffer(packet)}
}

/**
 * PacketWriter
 */

var truncateErrMessage string = "String was truncated to " + strconv.Itoa(math.MaxUint8) + "character(s)"

type PacketWriterErr struct {
	message string
}

func (err *PacketWriterErr) Error() string {
	if err.message != "" {
		return err.message
	} else {
		return "Failed to write one or more value(s)"
	}
}

type PacketWriter struct {
	buffer bytes.Buffer
}

func (writer *PacketWriter) Bytes() []byte {
	return writer.buffer.Bytes()
}

func (writer *PacketWriter) WriteU8(value uint8) error {
	return writer.buffer.WriteByte(value)
}

func (writer *PacketWriter) WriteU32(value uint32) error {
	return binary.Write(&writer.buffer, binary.LittleEndian, value)
}

func (writer *PacketWriter) WriteBool(value bool) {
	writer.WriteU8(util.BoolToU8(value))
}

func (writer *PacketWriter) WriteString(value string) bool {
	bytes := []byte(value)
	success := len(bytes) <= math.MaxUint8

	if !success {
		bytes = bytes[:math.MaxUint8+1]
	}

	writer.WriteU8(uint8(len(bytes)))
	writer.buffer.Write(bytes)

	return success
}

func (writer *PacketWriter) WriteU8Cond(value uint8, cond bool) {
	if cond {
		writer.WriteU8(value)
	}
}

func (writer *PacketWriter) WriteStringCond(value string, cond bool) {
	if cond {
		writer.WriteString(value)
	}
}

func (writer *PacketWriter) Write(values ...any) error {
	for _, value := range values {
		switch cast := value.(type) {
		case uint8:
			writer.WriteU8(cast)
		case uint32:
			writer.WriteU32(cast)
		case bool:
			writer.WriteBool(cast)
		case string:
			if !writer.WriteString(cast) {
				return &PacketWriterErr{message: truncateErrMessage}
			}
		default:
			return &PacketWriterErr{}
		}
	}

	return nil
}

func NewPacketWriter(size int) *PacketWriter {
	return &PacketWriter{}
}

File: client/public/js/packets/io.js
/**
 * Copyright (C) 2026 Ellie Fisher
 *
 * This file is part of the Word Magnets source code. It may be used under the GNU Affero General
 * Public License v3.0.
 *
 * For full terms, see the LICENSE file or visit https://spdx.org/licenses/AGPL-3.0-or-later.html
 */

import { clamp, enumerate } from "../util.js";

export const PacketTypes = enumerate([
	"InvalidPacket",

	/* Client=>Server */

	"RequestServerInfoPacket",
	"CreateRoomPacket",
	"JoinRoomPacket",
	"LeaveRoomPacket",
	"RemoveClientPacket",
	"StartGamePacket",
	"CancelStartGamePacket",
	"SubmitSentencePacket",
	"SubmitVotePacket",

	/* Server=>Client */

	"ClientInfoPacket",
	"ServerInfoPacket",
	"RoomConnectErrorPacket",
	"RoomDestroyedPacket",
	"RoomDataPacket",
	"RoomClientsPacket",
	"RoomWordsPacket",
	"RoomSentencesPacket",
]);

export const U8_MAX_VALUE = 255;

export const I8_MIN_VALUE = -128;
export const I8_MAX_VALUE = 127;

export const MAX_STRING_LENGTH = U8_MAX_VALUE;

/**
 * PacketReader is a class for reading Uint8Arrays.
 */
export class PacketReader {
	constructor(array) {
		this._array = array;
		this._view = new DataView(this._array.buffer);
		this._index = 0;
	}

	isAtEnd() {
		return this._index >= this._array.length;
	}

	readU8() {
		return this.isAtEnd() ? 0 : this._view.getUint8(this._index++);
	}

	readU32() {
		let value = 0;

		try {
			value = this._view.getUint32(this._index, true);
			this._index += 4;
		} catch {}

		return value;
	}

	readI8() {
		return this.isAtEnd() ? 0 : this._view.getInt8(this._index++);
	}

	readBool() {
		return this.readU8() != 0;
	}

	readString() {
		let str = "";
		const length = this.readU8();

		for (let i = 0; i < length; i++) {
			str += String.fromCharCode(this.readU8());
		}

		return str;
	}

	readU8Cond(outObject, key, test) {
		if (test) {
			outObject[key] = this.readU8();
		}
	}

	readStringCond(outObject, key, test) {
		if (test) {
			outObject[key] = this.readString();
		}
	}
}

/**
 * PacketWriter is a class for writing Uint8Arrays.
 */
export class PacketWriter {
	constructor() {
		this._array = [];
		this._index = 0;
	}

	bytes() {
		return new Uint8Array(this._array);
	}

	writeU8(value) {
		value = clamp(value, 0, U8_MAX_VALUE);

		if (this._index < this._array.length) {
			this._array[this._index] = value;
		} else {
			this._array.push(value);
		}

		this._index++;
	}

	writeI8(value) {
		value = clamp(value, I8_MIN_VALUE, I8_MAX_VALUE);

		if (this._index < this._array.length) {
			this._array[this._index] = value;
		} else {
			this._array.push(value);
		}

		this._index++;
	}

	writeBool(value) {
		this.writeU8(value != 0);
	}

	writeString(value) {
		const truncated = value.length > MAX_STRING_LENGTH;

		if (truncated) {
			value = value.slice(0, MAX_STRING_LENGTH);
		}

		this.writeU8(value.length);

		for (const ch of value) {
			this.writeU8(ch.charCodeAt(0));
		}

		return truncated;
	}

	write(...values) {
		let success = true;

		for (const value of values) {
			switch (typeof value) {
				case "number":
					success = Number.isInteger(value) && value >= 0 && value <= U8_MAX_VALUE;

					if (success) {
						this.writeU8(value);
					}

					break;

				case "string":
					this.writeString(value);
					break;

				default:
					success = false;
					break;
			}

			if (!success) {
				break;
			}
		}

		return success;
	}
}
